rust/stracciatella/src/vfs/android.rs

```diff
@@ -75,27 +93,42 @@ impl AssetManagerFs {
             if want == "." || want == ".." {
                 return Err(io::Error::new(
                     io::ErrorKind::InvalidInput,
-                    "special path components are not supported",
+                    "AssetManagerFs: Special path components are not supported",
                 ));
             }
             for candidate in candidates {
                 let candidate_cstring = Self::path_to_cstring(&candidate)?;
-                let dir_contents = crate::android::list_asset_dir(&candidate).map_err(|e| {
-                    io::Error::new(
-                        io::ErrorKind::Other,
-                        format!("AssetManagerFs: JNI Error: `{:?}`", e),
-                    )
-                })?;
-                if self.asset_manager.open_dir(&candidate_cstring).is_some() {
-                    for entry in dir_contents {
-                        let is_match = entry
-                            .file_name()
-                            .and_then(|x| x.to_str())
-                            .map(|x| want == Nfc::caseless(x).as_str())
-                            .unwrap_or(false);
-                        if is_match {
-                            next.push(candidate.join(&entry));
-                        }
+                let entries = if let Some(cache_entry) = canonicalization_cache.get(&candidate) {
+                    cache_entry
+                } else if self.asset_manager.open_dir(&candidate_cstring).is_none() {
+                    canonicalization_cache.put(candidate.clone(), vec![]);
+                    canonicalization_cache
+                        .get(&candidate)
+                        .expect("we should be able to get a cache key that was just set")
+                } else {
+                    let entries: Vec<_> = crate::android::list_asset_dir(&candidate)
+                        .map_err(|e| {
+                            io::Error::new(
+                                io::ErrorKind::Other,
+                                format!("AssetManagerFs: JNI Error: `{:?}`", e),
+                            )
+                        })?
+                        .iter()
+                        .flat_map(|path_buf| {
+                            path_buf
+                                .to_str()
+                                .map(|s| (Nfc::caseless(s), path_buf.into()))
+                        })
+                        .collect();
+                    canonicalization_cache.put(candidate.clone(), entries);
+                    canonicalization_cache
+                        .get(&candidate)
+                        .expect("we should be able to get a cache key that was just set")
+                };
+
+                for (nfc, os_string) in entries {
+                    if want == nfc.as_str() {
+                        next.push(candidate.join(&os_string));
```

Comment:
Can `entries` validly contain duplicates or can we continue here?

Reply:
In theory it can contain updates, though I am not sure if we need to handle it (e.g. to paths resolve to the same NFC). We could also just continue...

Reply:
I looked into it a little more. This case handles the issue when there are two paths e.g. `test/foo/baz` and `test/FOO/bar` that have the same base directory in different casing. Always choosing the first match here, would not allow us to find `test/FOO/bar` because `test/FOO` is always normalized to `test/foo`.